# Naive Bayes part 2 and 3 <a class="tocSkip">

In [1]:
import os
import io
import numpy 
from pandas import DataFrame
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
import pyearth

In [2]:
def readFiles(path):
    for root,dirnames,filenames in os.walk(path):
        for filename in filenames:
            path=os.path.join(root,filename)
            
            inBody=False
            lines =[]
            f = io.open(path,'r',encoding='latin1')
            for line in f:
                if inBody:
                    lines.append(line)
                elif line=='\n':
                    inBody=True
            f.close()
            message = '\n'.join(lines)
            yield path,message    

In [3]:
def dataFrameFromDirectory(path,classification):
    rows =[]
    index=[]
    for filename,message in readFiles(path):
        rows.append({'message': message,'class': classification})
        index.append(filename)
        
    return DataFrame(rows,index=index)

# Train set

In [4]:
train = DataFrame({'message':[],'class':[]})
train

Empty DataFrame
Columns: [message, class]
Index: []

In [5]:
train=train.append(dataFrameFromDirectory(r'C:\Users\Paul\Documents\_SXOLH\Machine Learning\Projects\Project_4\Project_4__classification\Project 4 - classification\emailspam\train\spam','spam'))
train=train.append(dataFrameFromDirectory(r'C:\Users\Paul\Documents\_SXOLH\Machine Learning\Projects\Project_4\Project_4__classification\Project 4 - classification\emailspam\train\ham','ham'))


C:\Users\Paul\.conda\envs\ml\lib\site-packages\pandas\core\frame.py:6692: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  sort=sort)


In [6]:
print(train.shape)
train.head()

(603, 2)


class  \
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   

                                                                                              message  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...                                                     
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  earn over $ 70 , 000 a month in your own home ...  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  this message complies with the proposed united...  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  make unlimited income no meetings , no phone c...  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  are you ready to reach new prospects without t...

In [7]:
#remove empty emails (there are two because of google drive .ini files)
train=train.loc[train['message']!='']
print(train.shape)
train.head()

(601, 2)


class  \
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   

                                                                                              message  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  earn over $ 70 , 000 a month in your own home ...  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  this message complies with the proposed united...  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  make unlimited income no meetings , no phone c...  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  are you ready to reach new prospects without t...  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  * * * * * * warning ! this message is for porn...

## Vectorizer

In [8]:
vectorizer =  CountVectorizer(min_df=1,lowercase=True,analyzer='word', binary=False, decode_error='strict',token_pattern='(?u)\\b\\w\\w+\\b')
x_train = vectorizer.fit_transform(train['message'].values)

In [9]:
trainfrequencies=DataFrame(x_train.toarray(),columns=vectorizer.get_feature_names())
trainfrequencies.head()

00  000  0000  0001  00014  000bp  001  00198  002656  0027  ...  \
0  21   14     0     0      0      0    0      0       0     0  ...   
1   0    0     0     0      0      0    0      0       0     0  ...   
2   0    0     0     0      0      0    0      0       0     0  ...   
3   0    0     0     0      0      0    0      0       0     0  ...   
4   0    0     0     0      0      0    0      0       0     0  ...   

   zukuenftigen  zulia  zulu  zur  zurich  zwart  zwarts  zweigenbaum  zwicky  \
0             0      0     0    0       0      0       0            0       0   
1             0      0     0    0       0      0       0            0       0   
2             0      0     0    0       0      0       0            0       0   
3             0      0     0    0       0      0       0            0       0   
4             0      0     0    0       0      0       0            0       0   

   zwischen  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 20939 columns]

In [10]:
y_train = train['class'].values

# Test set

In [11]:
test = DataFrame({'message':[],'class':[]})
test

Empty DataFrame
Columns: [message, class]
Index: []

In [12]:
test=test.append(dataFrameFromDirectory(r'C:\Users\Paul\Documents\_SXOLH\Machine Learning\Projects\Project_4\Project_4__classification\Project 4 - classification\emailspam\test\spam','spam'))
test=test.append(dataFrameFromDirectory(r'C:\Users\Paul\Documents\_SXOLH\Machine Learning\Projects\Project_4\Project_4__classification\Project 4 - classification\emailspam\test\ham','ham'))

In [13]:
print(test.shape)
test.head()

(203, 2)


class  \
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   

                                                                                              message  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...                                                     
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  if you want the best hunting and camping vacat...  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  57 million email addresses for only $ 99 you w...  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  attention ! warning ! adults only ! warning ! ...  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  attention ! warning ! adults only ! warning ! ...

In [14]:
#remove empty emails (there are two because of google drive .ini files)
test=test.loc[test['message']!='']
print(test.shape)
test.head()

(201, 2)


class  \
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  spam   

                                                                                              message  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  if you want the best hunting and camping vacat...  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  57 million email addresses for only $ 99 you w...  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  attention ! warning ! adults only ! warning ! ...  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  attention ! warning ! adults only ! warning ! ...  
C:\Users\Paul\Documents\_SXOLH\Machine Learning...  subject : re : are you in debt ? if you are th...

In [15]:
x_test = vectorizer.transform(test['message'].values)

In [16]:
testfrequencies=DataFrame(x_test.toarray(),columns=vectorizer.get_feature_names())
testfrequencies.head()

00  000  0000  0001  00014  000bp  001  00198  002656  0027  ...  \
0   1    0     0     0      0      0    0      0       0     0  ...   
1   4   10     0     0      0      0    0      0       0     0  ...   
2   1    0     0     0      0      0    0      0       0     0  ...   
3   1    0     0     0      0      0    0      0       0     0  ...   
4   1    1     0     0      0      0    0      0       0     0  ...   

   zukuenftigen  zulia  zulu  zur  zurich  zwart  zwarts  zweigenbaum  zwicky  \
0             0      0     0    0       0      0       0            0       0   
1             0      0     0    0       0      0       0            0       0   
2             0      0     0    0       0      0       0            0       0   
3             0      0     0    0       0      0       0            0       0   
4             0      0     0    0       0      0       0            0       0   

   zwischen  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 20939 columns]

In [17]:
y_test = test['class'].values

# Top T features

Word frequency dataframe with classes

In [18]:
#reset index 
dat1=DataFrame(train['class'])
dat1.columns=['mailclass']
dat2=trainfrequencies

dat1 = dat1.reset_index(drop=True)
dat2 = dat2.reset_index(drop=True)

trainfrequenciesClass=pd.concat([dat1,dat2],axis=1)
trainfrequenciesClass

mailclass  00  000  0000  0001  00014  000bp  001  00198  002656  ...  \
0        spam  21   14     0     0      0      0    0      0       0  ...   
1        spam   0    0     0     0      0      0    0      0       0  ...   
2        spam   0    0     0     0      0      0    0      0       0  ...   
3        spam   0    0     0     0      0      0    0      0       0  ...   
4        spam   0    0     0     0      0      0    0      0       0  ...   
5        spam   2    0     0     0      0      0    0      0       0  ...   
6        spam   0    0     0     0      0      0    0      0       0  ...   
7        spam   2    0     0     0      0      0    0      0       0  ...   
8        spam   2    0     0     0      0      0    0      0       0  ...   
9        spam   0    0     0     0      0      0    0      0       0  ...   
10       spam   0    0     0     0      0      0    0      0       0  ...   
11       spam   0    0     0     0      0      0    0      0       0  ...   
12       spam   0    0     0     0      0      0    0      0       0  ...   
13       spam   0    0     0     0      0      0    0      0       0  ...   
14       spam   4    0     0     0      0      0    0      0       0  ...   
15       spam   0   20     0     0      0      0    0      0       0  ...   
16       spam  11    0     0     0      0      0    0      0       0  ...   
17       spam   0    0     0     0      0      0    0      0       0  ...   
18       spam   0    0     0     0      0      0    0      0       0  ...   
19       spam   0    0     0     0      0      0    0      0       0  ...   
20       spam   0    0     0     0      0      0    0      0       0  ...   
21       spam   0    0     0     0      0      0    0      0       0  ...   
22       spam   0    0     0     0      0      0    0      0       0  ...   
23       spam   7   24     0     0      0      0    0      0       0  ...   
24       spam   0    0     0     0      0      0    0      0       0  ...   
25       spam   0    1     0     0      0      0    0      0       0  ...   
26       spam   1    0     0     0      0      0    0      0       0  ...   
27       spam   1   36     0     0      0      0    0      0       0  ...   
28       spam   0    0     0     0      0      0    0      0       0  ...   
29       spam   1    0     0     0      0      0    0      0       0  ...   
..        ...  ..  ...   ...   ...    ...    ...  ...    ...     ...  ...   
571       ham   1    0     0     0      0      0    0      0       0  ...   
572       ham   3    0     0     0      0      0    0      0       0  ...   
573       ham   0    0     0     0      0      0    0      0       0  ...   
574       ham  11    0     0     0      0      0    0      0       0  ...   
575       ham   0    0     0     0      0      0    0      0       0  ...   
576       ham   0    0     0     0      0      0    0      0       0  ...   
577       ham   0    0     0     0      0      0    0      0       0  ...   
578       ham   0    0     0     0      0      0    0      0       0  ...   
579       ham   0    0     0     0      0      0    0      0       0  ...   
580       ham   0    0     0     0      0      0    0      0       0  ...   
581       ham   0    0     0     0      0      0    0      0       0  ...   
582       ham   0    0     0     0      0      0    0      0       0  ...   
583       ham   0    0     0     0      0      0    0      0       0  ...   
584       ham   0    0     0     0      0      0    0      0       0  ...   
585       ham   0    0     0     0      0      0    0      0       0  ...   
586       ham   7    0     0     0      0      0    0      0       0  ...   
587       ham   0    0     0     0      0      0    0      0       0  ...   
588       ham   1    0     0     0      0      0    0      0       0  ...   
589       ham   0    0     0     0      0      0    0      0       0  ...   
590       ham   2    0     0     0      0      0    0      0       0  ...   


In [19]:
sums=DataFrame(columns=vectorizer.get_feature_names(),index=['count'])
sums.loc['count']=trainfrequenciesClass.sum(axis=0)[1:]
sums

00  000 0000 0001 00014 000bp 001 00198 002656 0027  ... zukuenftigen  \
count  728  840    1    2     3     1   1     4      1    1  ...            1   

      zulia zulu zur zurich zwart zwarts zweigenbaum zwicky zwischen  
count     1    1   2      2     1      3           1      4        1  

[1 rows x 20939 columns]

In [20]:
#more than 50 count
sums=sums[sums>50]
sums=sums[sums.columns[~sums.isnull().all()]]
sums

00  000   10  100   11   12  13  14   15   16  ... written  www year  \
count  728  840  559  344  170  156  70  93  255  104  ...      51  271  148   

      years yes yet   you  your yours yourself  
count   180  75  52  5592  2364    70       74  

[1 rows x 771 columns]

In [21]:
topwordslist=list(sums.columns)

#remove numbers
no_integers = [item for item in topwordslist if not item.isdigit()]
topwordslist=no_integers

topwordslist

['3d',
 'able',
 'about',
 'above',
 'abstract',
 'abstracts',
 'ac',
 'accepted',
 'access',
 'account',
 'acl',
 'acquisition',
 'action',
 'actually',
 'ad',
 'add',
 'additional',
 'address',
 'addresses',
 'ads',
 'adult',
 'advertise',
 'advertisement',
 'advertising',
 'after',
 'again',
 'against',
 'all',
 'allow',
 'allows',
 'almost',
 'along',
 'already',
 'also',
 'always',
 'am',
 'amazing',
 'american',
 'amount',
 'an',
 'analysis',
 'and',
 'another',
 'answer',
 'any',
 'anyone',
 'anything',
 'anywhere',
 'aol',
 'applications',
 'approach',
 'approaches',
 'april',
 'are',
 'area',
 'around',
 'artificial',
 'as',
 'ask',
 'asked',
 'aspects',
 'assume',
 'at',
 'august',
 'author',
 'authors',
 'available',
 'away',
 'back',
 'bank',
 'based',
 'basis',
 'be',
 'because',
 'become',
 'been',
 'before',
 'being',
 'believe',
 'below',
 'best',
 'better',
 'between',
 'big',
 'bills',
 'bonus',
 'book',
 'both',
 'box',
 'break',
 'bring',
 'build',
 'bulk',
 'busine

## Ham word sums

In [22]:
hamfrequencies=trainfrequenciesClass.loc[trainfrequenciesClass['mailclass']=='ham']
hamfrequencies.shape

(301, 20940)

In [23]:
hamsums=DataFrame(columns=vectorizer.get_feature_names(),index=['hamwordcount'])
hamsums.loc['hamwordcount']=hamfrequencies.sum(axis=0)[1:]
hamsums

00 000 0000 0001 00014 000bp 001 00198 002656 0027  ...  \
hamwordcount  277  41    0    2     3     1   1     4      1    1  ...   

             zukuenftigen zulia zulu zur zurich zwart zwarts zweigenbaum  \
hamwordcount            1     1    1   2      2     1      3           1   

             zwicky zwischen  
hamwordcount      4        1  

[1 rows x 20939 columns]

In [24]:
#how many emails included each word
hamsums.loc['hamemails']=hamfrequencies.astype(bool).sum(axis=0)
hamsums

00 000 0000 0001 00014 000bp 001 00198 002656 0027  ...  \
hamwordcount  277  41    0    2     3     1   1     4      1    1  ...   
hamemails      37  12    0    1     2     1   1     2      1    1  ...   

             zukuenftigen zulia zulu zur zurich zwart zwarts zweigenbaum  \
hamwordcount            1     1    1   2      2     1      3           1   
hamemails               1     1    1   2      2     1      3           1   

             zwicky zwischen  
hamwordcount      4        1  
hamemails         2        1  

[2 rows x 20939 columns]

In [25]:
#more than 50 count
topham = hamsums[hamsums.columns.intersection(topwordslist)]
topham

3d able about above abstract abstracts   ac accepted access  \
hamwordcount  0   28   242    58      104       101  136       59     38   
hamemails     0   27   113    44       56        47   51       36     18   

             account  ... written  www year years yes yet  you your yours  \
hamwordcount      41  ...      42  131   35    45  19  36  372  114     2   
hamemails         26  ...      30   58   24    25  12  26  114   50     2   

             yourself  
hamwordcount        2  
hamemails           1  

[2 rows x 731 columns]

## Spam word sums

In [26]:
spamfrequencies=trainfrequenciesClass.loc[trainfrequenciesClass['mailclass']=='spam']
spamfrequencies.shape

(300, 20940)

In [27]:
spamsums=DataFrame(columns=vectorizer.get_feature_names(),index=['spamwordcount'])
spamsums.loc['spamwordcount']=spamfrequencies.sum(axis=0)[1:]
spamsums

00  000 0000 0001 00014 000bp 001 00198 002656 0027  ...  \
spamwordcount  451  799    1    0     0     0   0     0      0    0  ...   

              zukuenftigen zulia zulu zur zurich zwart zwarts zweigenbaum  \
spamwordcount            0     0    0   0      0     0      0           0   

              zwicky zwischen  
spamwordcount      0        0  

[1 rows x 20939 columns]

In [28]:
#how many emails included each word
spamsums.loc['spamemails']=spamfrequencies.astype(bool).sum(axis=0)
spamsums

00  000 0000 0001 00014 000bp 001 00198 002656 0027  ...  \
spamwordcount  451  799    1    0     0     0   0     0      0    0  ...   
spamemails      94   93    1    0     0     0   0     0      0    0  ...   

              zukuenftigen zulia zulu zur zurich zwart zwarts zweigenbaum  \
spamwordcount            0     0    0   0      0     0      0           0   
spamemails               0     0    0   0      0     0      0           0   

              zwicky zwischen  
spamwordcount      0        0  
spamemails         0        0  

[2 rows x 20939 columns]

In [29]:
#more than 50 count
topspam = spamsums[spamsums.columns.intersection(topwordslist)]
topspam

3d able about above abstract abstracts  ac accepted access  \
spamwordcount  718   62   374    63        0         0  17       13     49   
spamemails      16   41   118    54        0         0   7       13     38   

              account  ... written  www year years yes yet   you  your yours  \
spamwordcount      59  ...       9  140  113   135  56  16  5220  2250    68   
spamemails         31  ...       9   79   57    64  42  13   264   221    48   

              yourself  
spamwordcount       72  
spamemails          55  

[2 rows x 731 columns]

## Joined in one dataframe

In [30]:
SpamandHamSums=pd.concat([topspam,topham],sort=False)
SpamandHamSums=SpamandHamSums.fillna(0)
SpamandHamSums

3d  able  about  above  abstract  abstracts   ac  accepted  \
spamwordcount  718    62    374     63         0          0   17        13   
spamemails      16    41    118     54         0          0    7        13   
hamwordcount     0    28    242     58       104        101  136        59   
hamemails        0    27    113     44        56         47   51        36   

               access  account  ...  written  www  year  years  yes  yet  \
spamwordcount      49       59  ...        9  140   113    135   56   16   
spamemails         38       31  ...        9   79    57     64   42   13   
hamwordcount       38       41  ...       42  131    35     45   19   36   
hamemails          18       26  ...       30   58    24     25   12   26   

                you  your  yours  yourself  
spamwordcount  5220  2250     68        72  
spamemails      264   221     48        55  
hamwordcount    372   114      2         2  
hamemails       114    50      2         1  

[4 rows x 731 columns]

In [31]:
totalwordcount=SpamandHamSums.loc['hamwordcount'] + SpamandHamSums.loc['spamwordcount']
totalemailcount=SpamandHamSums.loc['hamemails'] + SpamandHamSums.loc['spamemails']
SpamandHamSums.loc['totalWordCount']=totalwordcount
SpamandHamSums.loc['totalEmailCount']=totalemailcount
SpamandHamSums

3d  able  about  above  abstract  abstracts   ac  accepted  \
spamwordcount    718    62    374     63         0          0   17        13   
spamemails        16    41    118     54         0          0    7        13   
hamwordcount       0    28    242     58       104        101  136        59   
hamemails          0    27    113     44        56         47   51        36   
totalWordCount   718    90    616    121       104        101  153        72   
totalEmailCount   16    68    231     98        56         47   58        49   

                 access  account  ...  written  www  year  years  yes  yet  \
spamwordcount        49       59  ...        9  140   113    135   56   16   
spamemails           38       31  ...        9   79    57     64   42   13   
hamwordcount         38       41  ...       42  131    35     45   19   36   
hamemails            18       26  ...       30   58    24     25   12   26   
totalWordCount       87      100  ...       51  271   148    180   75   52   
totalEmailCount      56       57  ...       39  137    81     89   54   39   

                  you  your  yours  yourself  
spamwordcount    5220  2250     68        72  
spamemails        264   221     48        55  
hamwordcount      372   114      2         2  
hamemails         114    50      2         1  
totalWordCount   5592  2364     70        74  
totalEmailCount   378   271     50        56  

[6 rows x 731 columns]

In [32]:
eps=0.000000001 #to avoid divide by zero

pSpam = SpamandHamSums.loc['spamemails'] / trainfrequencies.shape[0]  #proportion of spam mails over all mails for each word
pHam = SpamandHamSums.loc['hamemails'] / trainfrequencies.shape[0]   #proportion of ham mails over all mails for each word

entropy = -pSpam*np.log2(pSpam + eps) - pHam*np.log2(pHam + eps)
entropy = np.round(entropy,decimals=2)
SpamandHamSums.loc['entropy']=entropy
SpamandHamSums

3d   able   about   above  abstract  abstracts      ac  \
spamwordcount    718.00  62.00  374.00   63.00      0.00       0.00   17.00   
spamemails        16.00  41.00  118.00   54.00      0.00       0.00    7.00   
hamwordcount       0.00  28.00  242.00   58.00    104.00     101.00  136.00   
hamemails          0.00  27.00  113.00   44.00     56.00      47.00   51.00   
totalWordCount   718.00  90.00  616.00  121.00    104.00     101.00  153.00   
totalEmailCount   16.00  68.00  231.00   98.00     56.00      47.00   58.00   
entropy            0.14   0.47    0.91    0.59      0.32       0.29    0.38   

                 accepted  access  account  ...  written     www    year  \
spamwordcount       13.00    49.0    59.00  ...     9.00  140.00  113.00   
spamemails          13.00    38.0    31.00  ...     9.00   79.00   57.00   
hamwordcount        59.00    38.0    41.00  ...    42.00  131.00   35.00   
hamemails           36.00    18.0    26.00  ...    30.00   58.00   24.00   
totalWordCount      72.00    87.0   100.00  ...    51.00  271.00  148.00   
totalEmailCount     49.00    56.0    57.00  ...    39.00  137.00   81.00   
entropy              0.36     0.4     0.42  ...     0.31    0.71    0.51   

                  years    yes    yet      you     your  yours  yourself  
spamwordcount    135.00  56.00  16.00  5220.00  2250.00  68.00     72.00  
spamemails        64.00  42.00  13.00   264.00   221.00  48.00     55.00  
hamwordcount      45.00  19.00  36.00   372.00   114.00   2.00      2.00  
hamemails         25.00  12.00  26.00   114.00    50.00   2.00      1.00  
totalWordCount   180.00  75.00  52.00  5592.00  2364.00  70.00     74.00  
totalEmailCount   89.00  54.00  39.00   378.00   271.00  50.00     56.00  
entropy            0.53   0.38   0.32     0.98     0.83   0.32      0.33  

[7 rows x 731 columns]

In [33]:
totalemailproportion=(SpamandHamSums.loc['spamemails'] + SpamandHamSums.loc['hamemails'])/trainfrequencies.shape[0]
totalemailproportion=np.round(totalemailproportion,decimals=2)
SpamandHamSums.loc['totalemailproportion']=totalemailproportion
SpamandHamSums

3d   able   about   above  abstract  abstracts  \
spamwordcount         718.00  62.00  374.00   63.00      0.00       0.00   
spamemails             16.00  41.00  118.00   54.00      0.00       0.00   
hamwordcount            0.00  28.00  242.00   58.00    104.00     101.00   
hamemails               0.00  27.00  113.00   44.00     56.00      47.00   
totalWordCount        718.00  90.00  616.00  121.00    104.00     101.00   
totalEmailCount        16.00  68.00  231.00   98.00     56.00      47.00   
entropy                 0.14   0.47    0.91    0.59      0.32       0.29   
totalemailproportion    0.03   0.11    0.38    0.16      0.09       0.08   

                          ac  accepted  access  account  ...  written     www  \
spamwordcount          17.00     13.00   49.00    59.00  ...     9.00  140.00   
spamemails              7.00     13.00   38.00    31.00  ...     9.00   79.00   
hamwordcount          136.00     59.00   38.00    41.00  ...    42.00  131.00   
hamemails              51.00     36.00   18.00    26.00  ...    30.00   58.00   
totalWordCount        153.00     72.00   87.00   100.00  ...    51.00  271.00   
totalEmailCount        58.00     49.00   56.00    57.00  ...    39.00  137.00   
entropy                 0.38      0.36    0.40     0.42  ...     0.31    0.71   
totalemailproportion    0.10      0.08    0.09     0.09  ...     0.06    0.23   

                        year   years    yes    yet      you     your  yours  \
spamwordcount         113.00  135.00  56.00  16.00  5220.00  2250.00  68.00   
spamemails             57.00   64.00  42.00  13.00   264.00   221.00  48.00   
hamwordcount           35.00   45.00  19.00  36.00   372.00   114.00   2.00   
hamemails              24.00   25.00  12.00  26.00   114.00    50.00   2.00   
totalWordCount        148.00  180.00  75.00  52.00  5592.00  2364.00  70.00   
totalEmailCount        81.00   89.00  54.00  39.00   378.00   271.00  50.00   
entropy                 0.51    0.53   0.38   0.32     0.98     0.83   0.32   
totalemailproportion    0.13    0.15   0.09   0.06     0.63     0.45   0.08   

                      yourself  
spamwordcount            72.00  
spamemails               55.00  
hamwordcount              2.00  
hamemails                 1.00  
totalWordCount           74.00  
totalEmailCount          56.00  
entropy                   0.33  
totalemailproportion      0.09  

[8 rows x 731 columns]

In [34]:
#we will select only words that are in at least 4 % of all e-mails.
#if a word appears in just a few mails ,it wont be that significant
indexes=np.where(SpamandHamSums.loc['totalemailproportion']>=0.04)[0]
SignificantWords=SpamandHamSums.iloc[:,indexes]
SignificantWords

able   about   above  abstract  abstracts      ac  \
spamwordcount         62.00  374.00   63.00      0.00       0.00   17.00   
spamemails            41.00  118.00   54.00      0.00       0.00    7.00   
hamwordcount          28.00  242.00   58.00    104.00     101.00  136.00   
hamemails             27.00  113.00   44.00     56.00      47.00   51.00   
totalWordCount        90.00  616.00  121.00    104.00     101.00  153.00   
totalEmailCount       68.00  231.00   98.00     56.00      47.00   58.00   
entropy                0.47    0.91    0.59      0.32       0.29    0.38   
totalemailproportion   0.11    0.38    0.16      0.09       0.08    0.10   

                      accepted  access  account  acquisition  ...  written  \
spamwordcount            13.00   49.00    59.00         0.00  ...     9.00   
spamemails               13.00   38.00    31.00         0.00  ...     9.00   
hamwordcount             59.00   38.00    41.00        89.00  ...    42.00   
hamemails                36.00   18.00    26.00        40.00  ...    30.00   
totalWordCount           72.00   87.00   100.00        89.00  ...    51.00   
totalEmailCount          49.00   56.00    57.00        40.00  ...    39.00   
entropy                   0.36    0.40     0.42         0.26  ...     0.31   
totalemailproportion      0.08    0.09     0.09         0.07  ...     0.06   

                         www    year   years    yes    yet      you     your  \
spamwordcount         140.00  113.00  135.00  56.00  16.00  5220.00  2250.00   
spamemails             79.00   57.00   64.00  42.00  13.00   264.00   221.00   
hamwordcount          131.00   35.00   45.00  19.00  36.00   372.00   114.00   
hamemails              58.00   24.00   25.00  12.00  26.00   114.00    50.00   
totalWordCount        271.00  148.00  180.00  75.00  52.00  5592.00  2364.00   
totalEmailCount       137.00   81.00   89.00  54.00  39.00   378.00   271.00   
entropy                 0.71    0.51    0.53   0.38   0.32     0.98     0.83   
totalemailproportion    0.23    0.13    0.15   0.09   0.06     0.63     0.45   

                      yours  yourself  
spamwordcount         68.00     72.00  
spamemails            48.00     55.00  
hamwordcount           2.00      2.00  
hamemails              2.00      1.00  
totalWordCount        70.00     74.00  
totalEmailCount       50.00     56.00  
entropy                0.32      0.33  
totalemailproportion   0.08      0.09  

[8 rows x 714 columns]

In [35]:
smartSignificance=(1-SignificantWords.loc['entropy'])*SignificantWords.loc['totalemailproportion']
smartSignificance=np.round(smartSignificance*100,decimals=2)
SignificantWords.loc['smartSignificance']=smartSignificance
SignificantWords

C:\Users\Paul\.conda\envs\ml\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


able   about   above  abstract  abstracts      ac  \
spamwordcount         62.00  374.00   63.00      0.00       0.00   17.00   
spamemails            41.00  118.00   54.00      0.00       0.00    7.00   
hamwordcount          28.00  242.00   58.00    104.00     101.00  136.00   
hamemails             27.00  113.00   44.00     56.00      47.00   51.00   
totalWordCount        90.00  616.00  121.00    104.00     101.00  153.00   
totalEmailCount       68.00  231.00   98.00     56.00      47.00   58.00   
entropy                0.47    0.91    0.59      0.32       0.29    0.38   
totalemailproportion   0.11    0.38    0.16      0.09       0.08    0.10   
smartSignificance      5.83    3.42    6.56      6.12       5.68    6.20   

                      accepted  access  account  acquisition  ...  written  \
spamwordcount            13.00   49.00    59.00         0.00  ...     9.00   
spamemails               13.00   38.00    31.00         0.00  ...     9.00   
hamwordcount             59.00   38.00    41.00        89.00  ...    42.00   
hamemails                36.00   18.00    26.00        40.00  ...    30.00   
totalWordCount           72.00   87.00   100.00        89.00  ...    51.00   
totalEmailCount          49.00   56.00    57.00        40.00  ...    39.00   
entropy                   0.36    0.40     0.42         0.26  ...     0.31   
totalemailproportion      0.08    0.09     0.09         0.07  ...     0.06   
smartSignificance         5.12    5.40     5.22         5.18  ...     4.14   

                         www    year   years    yes    yet      you     your  \
spamwordcount         140.00  113.00  135.00  56.00  16.00  5220.00  2250.00   
spamemails             79.00   57.00   64.00  42.00  13.00   264.00   221.00   
hamwordcount          131.00   35.00   45.00  19.00  36.00   372.00   114.00   
hamemails              58.00   24.00   25.00  12.00  26.00   114.00    50.00   
totalWordCount        271.00  148.00  180.00  75.00  52.00  5592.00  2364.00   
totalEmailCount       137.00   81.00   89.00  54.00  39.00   378.00   271.00   
entropy                 0.71    0.51    0.53   0.38   0.32     0.98     0.83   
totalemailproportion    0.23    0.13    0.15   0.09   0.06     0.63     0.45   
smartSignificance       6.67    6.37    7.05   5.58   4.08     1.26     7.65   

                      yours  yourself  
spamwordcount         68.00     72.00  
spamemails            48.00     55.00  
hamwordcount           2.00      2.00  
hamemails              2.00      1.00  
totalWordCount        70.00     74.00  
totalEmailCount       50.00     56.00  
entropy                0.32      0.33  
totalemailproportion   0.08      0.09  
smartSignificance      5.44      6.03  

[9 rows x 714 columns]

## Most significant words

In [36]:
#TOP 10
top=SignificantWords.loc['smartSignificance'].sort_values(ascending=False)[0:9]
top

language       12.90
linguistics    12.50
university     11.10
remove         10.45
free           10.23
languages       9.72
money           9.12
papers          9.12
click           8.85
Name: smartSignificance, dtype: float64

In [37]:
SignificantWords['click']

spamwordcount           157.00
spamemails               89.00
hamwordcount              0.00
hamemails                 0.00
totalWordCount          157.00
totalEmailCount          89.00
entropy                   0.41
totalemailproportion      0.15
smartSignificance         8.85
Name: click, dtype: float64

It actually looks like a decent measure in our opinion.

# Suggested features

We suggest three extra features.  

* Message size  
* Count of words 
* Occurrences of obvious spam words coming from an external dictionary.

In [38]:
calls_to_action="“act now“; “apply now“; “buy now“; “click below“; “click here“; “click me to download“; “click this link“; “click to remove“; “call free“; “call now“; “claim now“; “contact us immediately“; “get now“; “sign up free“; “shop now“; “order now“; “get paid”"
calls_to_action=calls_to_action.replace("“","")
calls_to_action=calls_to_action.replace("”","")
cta_list=calls_to_action.split(';')
cta_list

['act now',
 ' apply now',
 ' buy now',
 ' click below',
 ' click here',
 ' click me to download',
 ' click this link',
 ' click to remove',
 ' call free',
 ' call now',
 ' claim now',
 ' contact us immediately',
 ' get now',
 ' sign up free',
 ' shop now',
 ' order now',
 ' get paid']

In [39]:
money_and_finance="“$$$“; “£££“; “accounts“; “additional income“; “bank“; “bonus“; “cash“; “cost“; “credit“; “earn“; “earn $“; “earn money“; “earn per week“; “finance”; “financial advice“; “financial freedom”; “free investment”; “get your money”; “insurance“; “investment”; “investment advice”; “investment decision”; “invoice”, “lowest price”, “make £“, “make money“; “million“; “money“; “money back“; “nominated bank account“; “potential earnings“; “profit“; “refund“; “risk free“; “save“; “save $“; “stock alert“; “thousands“; “US Dollars”"
money_and_finance=money_and_finance.replace("“","")
money_and_finance=money_and_finance.replace("”","")
maf_list=money_and_finance.split(';')
maf_list

['$$$',
 ' £££',
 ' accounts',
 ' additional income',
 ' bank',
 ' bonus',
 ' cash',
 ' cost',
 ' credit',
 ' earn',
 ' earn $',
 ' earn money',
 ' earn per week',
 ' finance',
 ' financial advice',
 ' financial freedom',
 ' free investment',
 ' get your money',
 ' insurance',
 ' investment',
 ' investment advice',
 ' investment decision',
 ' invoice, lowest price, make £, make money',
 ' million',
 ' money',
 ' money back',
 ' nominated bank account',
 ' potential earnings',
 ' profit',
 ' refund',
 ' risk free',
 ' save',
 ' save $',
 ' stock alert',
 ' thousands',
 ' US Dollars']

In [40]:
marketing="“ad“; “amazing“; “bargain“; “beneficial offer“; “cheap“; “clearance“; “congratulations“; “”dear“; “direct marketing“; “don’t delete“; “email marketing“; “fantastic“; “free“; “free trial” (+ many other phrases containing free, so please avoid); “gift certificate“; “increase sales“; “increase traffic“; “internet marketing“; “junk“; “marketing“; “marketing solution“; “member“; “message from“; “month trial offer“; “off everything“; “offer“; “offer expires“; “offer extended“; “online marketing“; “opportunity“; “opt in“; “performance“; “promise you“; “sale“; “search engine optimisation“; “spam“; “special promotion”; “stop further distribution“; “subscribe“; “super promo“; “the following form“; “this isn’t junk“; “this isn’t spam“; “top urgent“; “unbeatable offer“; “unsubscribe“; “urgent“; “urgent response“; “visit our website“; “web traffic“; “win“; “winner”"
marketing=marketing.replace("“","")
marketing=marketing.replace("”","")
m_list=marketing.split(';')
m_list

['ad',
 ' amazing',
 ' bargain',
 ' beneficial offer',
 ' cheap',
 ' clearance',
 ' congratulations',
 ' dear',
 ' direct marketing',
 ' don’t delete',
 ' email marketing',
 ' fantastic',
 ' free',
 ' free trial (+ many other phrases containing free, so please avoid)',
 ' gift certificate',
 ' increase sales',
 ' increase traffic',
 ' internet marketing',
 ' junk',
 ' marketing',
 ' marketing solution',
 ' member',
 ' message from',
 ' month trial offer',
 ' off everything',
 ' offer',
 ' offer expires',
 ' offer extended',
 ' online marketing',
 ' opportunity',
 ' opt in',
 ' performance',
 ' promise you',
 ' sale',
 ' search engine optimisation',
 ' spam',
 ' special promotion',
 ' stop further distribution',
 ' subscribe',
 ' super promo',
 ' the following form',
 ' this isn’t junk',
 ' this isn’t spam',
 ' top urgent',
 ' unbeatable offer',
 ' unsubscribe',
 ' urgent',
 ' urgent response',
 ' visit our website',
 ' web traffic',
 ' win',
 ' winner']

In [41]:
online_bussiness="“additional income“; “be your own boss“; “extra income“; “free hosting“; “get paid“; “home based business“; “home employment“; “income from home“; “profit“; “sale“; “work at home“; “work from home“; “while you sleep”"
online_bussiness=online_bussiness.replace("“","")
online_bussiness=online_bussiness.replace("”","")
ob_list=online_bussiness.split(';')
ob_list

['additional income',
 ' be your own boss',
 ' extra income',
 ' free hosting',
 ' get paid',
 ' home based business',
 ' home employment',
 ' income from home',
 ' profit',
 ' sale',
 ' work at home',
 ' work from home',
 ' while you sleep']

In [42]:
spam_word_list= cta_list + maf_list + m_list + ob_list
spam_word_list

['act now',
 ' apply now',
 ' buy now',
 ' click below',
 ' click here',
 ' click me to download',
 ' click this link',
 ' click to remove',
 ' call free',
 ' call now',
 ' claim now',
 ' contact us immediately',
 ' get now',
 ' sign up free',
 ' shop now',
 ' order now',
 ' get paid',
 '$$$',
 ' £££',
 ' accounts',
 ' additional income',
 ' bank',
 ' bonus',
 ' cash',
 ' cost',
 ' credit',
 ' earn',
 ' earn $',
 ' earn money',
 ' earn per week',
 ' finance',
 ' financial advice',
 ' financial freedom',
 ' free investment',
 ' get your money',
 ' insurance',
 ' investment',
 ' investment advice',
 ' investment decision',
 ' invoice, lowest price, make £, make money',
 ' million',
 ' money',
 ' money back',
 ' nominated bank account',
 ' potential earnings',
 ' profit',
 ' refund',
 ' risk free',
 ' save',
 ' save $',
 ' stock alert',
 ' thousands',
 ' US Dollars',
 'ad',
 ' amazing',
 ' bargain',
 ' beneficial offer',
 ' cheap',
 ' clearance',
 ' congratulations',
 ' dear',
 ' direct m

## Train

In [43]:
import sys

#obvious spam word count , count of words and size of message feature creation

obvious_spam_train=np.zeros(train['message'].shape[0])
count_of_words_train=np.zeros(train['message'].shape[0])
size_of_message_train=np.zeros(train['message'].shape[0])

for i,message in zip(range(train['message'].shape[0]),train['message']):
    totaloccurrences=0
    for word in spam_word_list:
        if word in message:
            occurrences=message.count(word)
            #print(word,occurrences)
            totaloccurrences+=occurrences
     
    obvious_spam_train[i]=totaloccurrences
    count_of_words_train[i]=len(message.split())
    size_of_message_train[i]=sys.getsizeof(message)

## Test

In [44]:
#obvious spam word count , count of words and size of message feature creation

obvious_spam_test=np.zeros(test['message'].shape[0])
count_of_words_test=np.zeros(test['message'].shape[0])
size_of_message_test=np.zeros(test['message'].shape[0])

for i,message in zip(range(test['message'].shape[0]),test['message']):
    totaloccurrences=0
    for word in spam_word_list:
        if word in message:
            occurrences=message.count(word)
            #print(word,occurrences)
            totaloccurrences+=occurrences
     
    obvious_spam_test[i]=totaloccurrences
    count_of_words_test[i]=len(message.split())
    size_of_message_test[i]=sys.getsizeof(message)

## kNN Accuracy with suggested features

In [45]:
#TRY WITH NEW FEATURES.OLD ACC WAS 0.965

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score


T=20

#TOP T FEATURES
top=SignificantWords.loc['smartSignificance'].sort_values(ascending=False)[0:T]
newtestfrequencies=testfrequencies[top.keys()]
newtrainfrequencies=trainfrequencies[top.keys()]


newtrainfrequencies.insert(loc=0,column='obvious_spam',value=obvious_spam_train)
newtestfrequencies.insert(loc=0,column='obvious_spam',value=obvious_spam_test)

#newtrainfrequencies.insert(loc=0,column='count of words',value=count_of_words_train)
#newtestfrequencies.insert(loc=0,column='count of words',value=count_of_words_test)

#newtrainfrequencies.insert(loc=0,column='message size',value=size_of_message_train)
#newtestfrequencies.insert(loc=0,column='message size',value=size_of_message_test)


#x_train  y_train
message_train=np.array(newtrainfrequencies)
spam_nospam_train=y_train

#x_test  y_test
message_test=np.array(newtestfrequencies)
spam_nospam_test=y_test


#KNN model
Spam_model = KNeighborsClassifier(n_neighbors=3)
Spam_model.fit(message_train, spam_nospam_train)

pred = Spam_model.predict(message_test)
accuracy_score(spam_nospam_test,pred)

0.9303482587064676

Accuracy seems to have dropped compared to kNN sklearn implementation in Problem 4.

# Naive Bayes

## Using the entire dataset

In [46]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score

classifier = MultinomialNB()
classifier.fit(x_train,y_train)

predictions=classifier.predict(x_test)

naive_bayes_accuracy_score=accuracy_score(y_test,predictions)
naive_bayes_accuracy_score

1.0

## Random subsets training (20%, 40%, 60%, 80%, and 100%)

In [94]:
import random

for portion in [0.2,0.4,0.6,0.8,1.0]:
    scores=[]

    #5 runs
    for i in range(5):

        row_count=int(portion*x_train.shape[0])

        #get random indices
        indices=list(np.arange(x_train.shape[0]))
        shuffled_indices = random.sample(indices, len(indices))
        final_indices=shuffled_indices[0:row_count]

        final_x_train = x_train[final_indices,:] #random subset of portion of x_train
        final_y_train = y_train[final_indices]

        #train Naive Bayes model
        classifier = MultinomialNB()
        classifier.fit(final_x_train,final_y_train)

        predictions=classifier.predict(x_test)

        naive_bayes_accuracy_score=accuracy_score(y_test,predictions)

        scores.append(naive_bayes_accuracy_score)

    average_score=np.average(scores)
    print("Average score over 5 runs using",round(portion*100),"percent of the training data:",round(average_score,ndigits=3))

Average score over 5 runs using 20 percent of the training data: 0.993
Average score over 5 runs using 40 percent of the training data: 0.997
Average score over 5 runs using 60 percent of the training data: 0.995
Average score over 5 runs using 80 percent of the training data: 0.997
Average score over 5 runs using 100 percent of the training data: 1.0


We can see that the Naive Bayes classifier **works great even when its trained using less data**.There is a small increase in performance with more data. We saw great scores even when it was trained with very small portions of the training data (1-5 percent) but they were not so consistent when picked randomly.

## Cross-validation

In [270]:
from random import randrange
 
       
# Split a dataset into k folds
def cross_validation_split(dataset, folds=3):
    dataset_split = list()
    dataset_copy = list(dataset)
    fold_size = int(len(dataset) / folds)
    for i in range(folds):
        fold = list()
        while len(fold) < fold_size:
            index = randrange(len(dataset_copy))
            fold.append(dataset_copy.pop(index))
        dataset_split.append(fold)
    return dataset_split

### Create concatenated dataset (xtrain + ytrain)

In [295]:
xdata=x_train.toarray()
print('xdata:')
#print(xdata)
print(xdata.shape)
print('\n')

y_data=y_train
print('y_data:')
#print(y_data)
print(y_data.shape)
print('\n')


alldata=np.c_[xdata,y_data]
print("ALL DATA\n")    
#print(alldata)
print(alldata.shape)

xdata:
(601, 20939)


y_data:
(601,)


ALL DATA

(601, 20940)


### Choose value of k for k-fold cross validation

In [297]:
number_folds=10  #choose value of k

number_rows=xdata.shape[0]
fold_size = round(number_rows / number_folds)
print('Fold size:',fold_size)

Fold size: 60


### Training and testing on different folds

In [303]:
folds = cross_validation_split(alldata, folds=number_folds)

accuracies=[]

for i in range(number_folds):
     
    #------TEST data------
    print('\n Testing on fold',i)
    testfold=np.array(folds[i])
    #print('Test fold:\n',testfold)
    
    
    
    #------TRAIN data------
    #remove test fold index
    trainindices=np.arange(number_folds).tolist()
    trainindices.pop(i)
    print('Training on folds',trainindices)
    
    #print("Train dataset:")
    train_data=np.zeros((0,alldata.shape[1])) 
    for j in trainindices:
        train_data=np.vstack((train_data,folds[j]))
    #print(train_data)
    
    
    
    
    #train and test data is ready but its the x and y data concatenated.we must split them  
    final_x_train=train_data[:,0:(train_data.shape[1]-1)]
    final_y_train=train_data[:,(train_data.shape[1]-1)]
    
    
    
    
    #train Naive Bayes model
    classifier = MultinomialNB()
    classifier.fit(final_x_train,final_y_train)

    predictions=classifier.predict(x_test)

    naive_bayes_accuracy_score=accuracy_score(y_test,predictions)

    print('ACCURACY:',naive_bayes_accuracy_score)
    
    accuracies.append(naive_bayes_accuracy_score)
    


 Testing on fold 0
Training on folds [1, 2, 3, 4, 5, 6, 7, 8, 9]
ACCURACY: 0.9950248756218906

 Testing on fold 1
Training on folds [0, 2, 3, 4, 5, 6, 7, 8, 9]
ACCURACY: 1.0

 Testing on fold 2
Training on folds [0, 1, 3, 4, 5, 6, 7, 8, 9]
ACCURACY: 1.0

 Testing on fold 3
Training on folds [0, 1, 2, 4, 5, 6, 7, 8, 9]
ACCURACY: 0.9950248756218906

 Testing on fold 4
Training on folds [0, 1, 2, 3, 5, 6, 7, 8, 9]
ACCURACY: 0.9900497512437811

 Testing on fold 5
Training on folds [0, 1, 2, 3, 4, 6, 7, 8, 9]
ACCURACY: 0.9950248756218906

 Testing on fold 6
Training on folds [0, 1, 2, 3, 4, 5, 7, 8, 9]
ACCURACY: 1.0

 Testing on fold 7
Training on folds [0, 1, 2, 3, 4, 5, 6, 8, 9]
ACCURACY: 0.9950248756218906

 Testing on fold 8
Training on folds [0, 1, 2, 3, 4, 5, 6, 7, 9]
ACCURACY: 1.0

 Testing on fold 9
Training on folds [0, 1, 2, 3, 4, 5, 6, 7, 8]
ACCURACY: 1.0


In [304]:
average_accuracy=np.average(accuracies)
average_accuracy

0.9970149253731343

The cross-validation results seem pretty **reliable** ,compared to evaluating accuracy on a separate test set.  
This however, is an extreme case where the accuracy is extremely good and it would be interesting to see what are the deviations from the test set accuracy, when the model is not so good.  
Still though, cross-validation is a very popular and widely used model validation technique.

# Feature selection effects

In [47]:
for T in [10,20,50,100,200,500]:
    
    #TOP T FEATURES
    top=SignificantWords.loc['smartSignificance'].sort_values(ascending=False)[0:T]

    newtestfrequencies=testfrequencies[top.keys()]
    newtrainfrequencies=trainfrequencies[top.keys()]
     
    #x_train  y_train
    x_train=np.array(newtrainfrequencies)
    y_train=y_train #just for aesthetics

    #x_test  y_test
    x_test=np.array(newtestfrequencies)
    y_test=y_test #just for aesthetics
        
        
    #Naive Bayes
    classifier = MultinomialNB()
    classifier.fit(x_train,y_train)
    predictions=classifier.predict(x_test)
    naive_bayes_accuracy_score=accuracy_score(y_test,predictions)
                
    print('\nUsing ',T,'most significant words\n')    
    print("Test set accuracy score: ",naive_bayes_accuracy_score)    


Using  10 most significant words

Test set accuracy score:  0.9353233830845771

Using  20 most significant words

Test set accuracy score:  0.9751243781094527

Using  50 most significant words

Test set accuracy score:  0.9900497512437811

Using  100 most significant words

Test set accuracy score:  0.9900497512437811

Using  200 most significant words

Test set accuracy score:  0.9950248756218906

Using  500 most significant words

Test set accuracy score:  0.9950248756218906


Naive Bayes seems to work **better with more features** and there aren't any dimensionality problems.We get almost perfect accuracy when using the top 50 most significant words or more.

When using the entire dataset (like we did above on 5.1), we get an accuracy of 1.0 which is amazing.